benchmark/suite.js

```diff
@@ -0,0 +1,126 @@
+import vm from 'node:vm';
+import benchmark from 'benchmark';
+import psl from '../index.js';
+
+const runSuite = ({ module, version }) => new Promise((resolve, reject) => {
+  const suite = new benchmark.Suite(version);
+
+  suite.add('psl#isValid', {
+    fn: () => {
+      module.isValid('google.com');
+    }
+  });
+
+  suite.add('psl#parse', {
+    fn: () => {
+      module.parse('google.com');
+    }
+  });
+
+  suite.add('psl#parse invalid domain', {
+    fn: () => {
+      module.parse('google.comp');
+    }
+  });
+
+  suite.on('error', reject);
+
+  suite.on('complete', () => {
+    resolve(suite.map(test => ({ version, ...test })));
+  });
+
+  suite.run({ async: true });
+});
+
+const runSuites = (suites) => {
+  const recurse = async (remaining, memo) => {
+    if (!remaining.length) {
+      return memo;
+    }
+
+    return recurse(remaining.slice(1), memo.concat(await runSuite(remaining[0])));
+  };
+
+  return recurse(suites, [])
+};
+
+const printResults = (results, compareToVersion) => {
+  const parsedResults = results.map(result => ({
+    version: result.version,
+    name: result.name,
+    'ops/sec': parseInt(result.hz.toFixed(result.hz < 100 ? 2 : 0), 10),
+    '\xb1 %': parseFloat(result.stats.rme.toFixed(2), 10),
+    'runs sampled': result.stats.sample.length,
+  }));
+
+  console.table(parsedResults.reduce(
+    (memo, { version, name, ...rest }) => ({
+      ...memo,
+      [`[${version}] ${name}`]: rest,
+    }),
+    {},
+  ));
+
+  const resultsByFunc = parsedResults.reduce(
+    (memo, result) => ({
+      ...memo,
+      [result.name]: {
+        ...memo[result.name],
+        [result.version]: result['ops/sec'],
+      },
+    }),
+    {},
+  );
+
+  const resultsByFuncSummary = Object.keys(resultsByFunc).reduce(
+    (memo, name) => {
+      const sourceOpsXSec = parseInt(resultsByFunc[name].source, 10);
+      const compareToOpsXSec = parseInt(resultsByFunc[name][compareToVersion], 10);
+      const diff = (
+        sourceOpsXSec > compareToOpsXSec
+          ? `${(sourceOpsXSec / compareToOpsXSec).toFixed(2)}x up`
+          : `${(compareToOpsXSec / sourceOpsXSec).toFixed(2)}x down`
+      );
+
+      return memo.concat({
+        name,
+        ...resultsByFunc[name],
+        diff,
+      });
+    },
+    [],
+  );
+
+  console.table(resultsByFuncSummary.reduce(
+    (memo, { name, ...rest }) => ({ ...memo, [name]: rest }),
+    {},
+  ));
+};
+
+const fetchModule = async (version) => {
+  const url = `https://unpkg.com/psl@${version}/dist/psl.mjs`;
+  const response = await fetch(url);
+  const text = await response.text();
+  const mod = new vm.SourceTextModule(text);
+
+  await mod.link(() => { });
+  await mod.evaluate();
+
+  return mod.namespace;
+};
+
+const main = async () => {
+  const compareToVersion = 'v1.13.0';
```

Comment:
I tried my best not to let this bother me. :rofl: 

I understand it allows us to test to different versions, in this case this would be the same as just having `url` be `https://unpkg.com/psl@latest/dist/psl.mjs` but we might want to test other cases, I get it.

I guess I should first ask how are we thinking about treating this file, is it something we're going to be updating on every release? are we always going to be testing against the same value we could find as `version` on our `package.json`? Or maybe I'm reading too much into it :shrug: idk, I'm trying to think of a way to keep this without having a human remembering to change this value.

Reply:
So far this script has only been used for manually benchmarking against the precending version (the current stable release) in the context of this PR (which is specifically about performance, and the original PR #302 was responsible for adding the script). In the future we could automate the benchmarking to always automatically compare new changes to the _current_ published version, but would only look into it when and if we feel it would add value (when addressing performance issues)..

Reply:
It could be a good idea to pass the version we want to compare to as a command line argument. That way, by default it would compare the local source code vs the _latest_ published version.

```sh
npm run benchmark
```

Alternatively, you would also be able to compare to a specific published version (ie: `v1.13.0`):

```sh
npm run benchmark -- v1.13.0
```

To achieve this I think we'd only need to change the following line:

https://github.com/lupomontero/psl/blob/440248a70c5c42aaa87a9b9cca349f506d367482/benchmark/suite.js#L113

Into this:

```js
const compareToVersion = process.argv[2] || 'latest';
```

Reply:
I've created a separate issue for this so we can proceed with this PR before adding more changes. See #335